torch/csrc/jit/codegen/cuda/glfdc/eval.h

```diff
@@ -0,0 +1,182 @@
+#pragma once
+
+// Expression evaluator
+
+#include <torch/csrc/jit/codegen/cuda/glfdc/bitvector.h>
+#include <torch/csrc/jit/codegen/cuda/glfdc/expr.h>
+#include <torch/csrc/jit/codegen/cuda/glfdc/sexpr.h>
+#include <torch/csrc/jit/codegen/cuda/glfdc/sparse_map.h>
+#include <torch/csrc/jit/codegen/cuda/glfdc/stack.h>
+
+#include <algorithm>
+#include <functional>
+
+#include <c10/util/Optional.h>
+
+namespace torch {
+namespace jit {
+namespace fuser {
+namespace cuda {
+namespace glfdc {
+
+struct Expr;
+struct ExprDAG;
+
+struct EvalState;
+
+// Operation for evaluator stack machine
+struct Operation {
+  const SExprRef ref;
+
+  // TODO: bitfields/ushort? 12b/16b should be plenty and we could pack it more
+  unsigned nsubops; // number of following operations to skip if lazy evaluated
+                    // (makes possible to recover tree structure)
+  unsigned noperands; // total number of operands on stack to calculate
+                      // subexpression value
+
+  static constexpr unsigned INVALID_POP = unsigned(-1);
+};
+
+using opt_index_t = c10::optional<std::size_t>;
+
+// Holds mapping of reused subexpressions
+struct ReusedExprMapping {
+  // Mapping of memoized values to slot values
+  const sparse_map slot_mapping;
+
+  // Given subexpressions returns its slot number or nullopt
+  opt_index_t slot(SExprRef ref) const noexcept {
+    return slot_mapping.find(ref.index_);
+  }
+
+  std::size_t size() const noexcept {
+    return slot_mapping.size();
+  }
+
+  bool empty() const noexcept {
+    return slot_mapping.empty();
+  }
+
+  // Creates empty mapping for eager evaluation of expression
+  static ReusedExprMapping createEagerMapping() {
+    return {sparse_map{}};
+  }
+
+  // create_lazy_mapping - Allocates mapping for reused subexpression and its
+  // value slot index
+  static ReusedExprMapping createLazyMapping(
+      const BitVector& reused_sexprs) // O(n)
+  {
```

Comment:
The convention here places `{` directly after the parameter declaration without a newline. I think clang-format gets confused with the comment (`// O(n)`) in this particular case.